langkit/templates/pkg_rewriting_impl_body_ada.mako

```diff
@@ -27,136 +27,120 @@ package body ${ada_lib_name}.Rewriting_Implementation is
-
-   procedure Pre_Check_Null_Or_Untied
-     (Handle     : Node_Rewriting_Handle) is
-   begin
-      if Handle /= No_Node_Rewriting_Handle and then Tied (Handle) then
-         Pre_Fail ("The given node must not be tied to another " &
-                   "rewriting context.");
-      end if;
-   end Pre_Check_Null_Or_Untied;
-   -- Check that the given node rewriting handle is not tied to any unit
-   -- rewriting handle.
-   -- Note: Null matches this predicate.
+   ---------------
+   -- PRE_CHECK --
+   ---------------
 
-   procedure Pre_Check_Tied
-     (Handle     : Node_Rewriting_Handle) is
-   begin
-      if not Tied (Handle) then
-         Pre_Fail ("The given node must be tied to an analysis unit.");
-      end if;
-   end Pre_Check_Tied;
-   -- Check that the given node rewriting handle is well tied to an analysis
-   -- unit.
+   procedure Pre_Check (Value : Boolean; Msg : String);
+   --  Raise a Precondition_Failure exception with the given message
+   --  if the Value is False.
 
-   procedure Pre_Check_Null_Or_Valid_Context
-     (Node : Node_Rewriting_Handle;
-      Handle : Rewriting_Handle) is
+   procedure Pre_Check (Value : Boolean; Msg : String) is
    begin
-      if Node /= No_Node_Rewriting_Handle and then Context (Node) = Handle then
-         Pre_Fail ("The given node has an invalid context.");
+      if not Value then
+         raise Precondition_Failure with Msg;
       end if;
-   end Pre_Check_Null_Or_Valid_Context;
-   -- Check that the node rewriting handle is either null, or belongs to the
-   -- given rewriting handle.
+   end Pre_Check;
+
+   <%def name="pre_check_rw_handle(handle)">
+      --  Check that rewriting handle ${handle} is not null
+      Pre_Check
+        (${handle} /= No_Rewriting_Handle,
+         "${handle} should not be null");
+   </%def>
+
+   <%def name="pre_check_rw_no_handle(handle)">
+      --  Check that rewriting handle ${handle} *is* null
+      Pre_Check
+        (${handle} = No_Rewriting_Handle,
+         "${handle} must be null");
+   </%def>
+
+   <%def name="pre_check_urw_handle(handle)">
+      --  Check that unit rewriting handle ${handle} is not null
+      Pre_Check
+        (${handle} /= No_Unit_Rewriting_Handle,
+         "${handle} should not be null");
+   </%def>
+
+   <%def name="pre_check_nrw_handle(handle)">
+      --  Check that node rewriting handle ${handle} is not null
+      Pre_Check
+        (${handle} /= No_Node_Rewriting_Handle,
+         "${handle} should not be null");
+   </%def>
+
+   <%def name="pre_check_child_index(handle, index, for_insertion)">
+      <%
+      upper_bound = "Children_Count ({}){}".format(
+         handle,
+         " + 1" if for_insertion else ""
+      )
+      %>
+      --  Check that ${index} is in range 1 .. ${upper_bound}.
+      Pre_Check
+        (${index} <= ${upper_bound},
+         "Invalid index " & ${index}'Image & ": ${handle} has " &
+         Children_Count (${handle})'Image & " children");
```

Comment:
This string formatting will happen everytime the corresponding check is performed, looks wasteful. I suggest to remove the dynamic computation in order to have a static string: `Out of bound child index` seems enough for the caller to understand what’s wrong.

Or even better: inline Pre_Check, so that the string formatting happens iff there is a precondition failure.